pkg/internal/controller/controller.go

```diff
@@ -383,6 +441,15 @@ func (c *Controller[request]) updateMetrics(reconcileTime time.Duration) {
 	ctrlmetrics.ReconcileTime.WithLabelValues(c.Name).Observe(reconcileTime.Seconds())
 }
 
+// ensureDidWarmupFinishChanInitialized ensures that the didWarmupFinishChan is initialized. This is needed
+// because controller can directly be created from other packages like controller.Controller, and
+// there is no way for the caller to pass in the chan.
+func (c *Controller[request]) ensureDidWarmupFinishChanInitialized() {
```

Comment:
Please instead write a `New` func in pkg/internal/controller and do it there so we don't need this code and the locking around it

Reply:
Good point, didn't realize it was in `internal` so was worried about keeping backwards compatibility

Reply:
c9a2973

Reply:
The reason I did not make New return some sort of interface and keep Controller private is because it would create a circular dep between `pkg/internal/controller` and `pkg/controller`.
I would have had to either define a new interface that replicates the ones in `/pkg/controller/controller.go` which adds the burden of keeping them in sync, so think this warning is good enough given that it's an internal pkg?